pym/bob/input.py

```diff
@@ -2545,7 +2555,7 @@ def prepare(self, inputEnv, sandboxEnabled, inputStates, inputSandbox=None,
         # optional checkout step
         if self.__checkout != (None, None, None) or self.__checkoutSCMs or self.__checkoutAsserts:
             checkoutDigestEnv = env.prune(self.__checkoutVars)
-            checkoutEnv = ( env.prune(self.__checkoutVars | self.__checkoutVarsWeak)
+            checkoutEnv = ( env.prune(self.__checkoutVars | self.__checkoutVarsWeak | depVars)
```

Comment:
This can't be right. Why are the substituted variables implicitly added to the environment variables? If the user wants to access these variables then he can add easily add them to `checkoutVars`...

Reply:
Hmm...the reason is most likely my misunderstanding of the code. I guess it must be added to `[checkout|build|package]DigestEnv` only to get different `[checkout|build|package]Ids`?

But for the `checkoutId` this is only necessary if `checkoutDep` is set, isn't it?

Reply:
> Hmm...the reason is most likely my misunderstanding of the code. I guess it must be added to `[checkout|build|package]DigestEnv` only to get different `[checkout|build|package]Ids`?

No, this is not needed. You can safely remove this and it should work as expected.

Reply:
I'm lost. If I remove the `| depVars` changes the unit `test_input_recipeset` is failing Because the package is always the same regardless of `$A` setting... 🤔

Reply:
This points to some yet not understood problem. I don't have an idea at the moment but I'm looking into it...

Reply:
Actually this is nothing new. The package graph is aggressively consolidated based on the following premise: if a particular recipe is fed with different environment+tools but yields the same result (as in Variant-Id) then the package graph is collapsed to a single variant. Usually you don't notice this but in your testcase the `b-foo` and `c-foo` recipes are identical so this optimization kicks in. You can get the same effect if you change `a.yaml` to:

```yaml
depends:
    - name: b-foo
      if: "$(eq,$A,b)"
    - name: c-foo
      if: "$(eq,$A,c)"
buildScript: "true"
packageScript: "true"
provideDeps: [ "*" ]
provideVars:
    D: "e"
```

So it's better to fix the test case by making the scripts in  `b-foo` and `c-foo` different.

Reply:
Hmm...so whenever we use a variable only in the depends section we'll get potentially wrong results? I wonder why I never fall about this.
My usecase is to provide different configs to busybox:

```yaml
depends
 - "${BUSYBOX_CONFIG:-core::busybox-config}"

buildScript: |
  ...
  if  [[ -e $2/busybox.config ]]; then  
     cp $2/busybox.config .
  fi
```

-> There is no obvious need to add `BUSYBOX_CONFIG` as `buildVar`. I think this should be at least documented somewhere...

Reply:
> Hmm...so whenever we use a variable only in the depends section we'll get potentially wrong results?

No, Bob will do the right thing **content wise**. If your `$BUSYBOX_CONFIG` points to different recipes that provide *different* configurations (technically: the variant-id of the packages must differ) then everything will be as expected. The only unexpected thing that happens is when `$BUSYBOX_CONFIG` points to different recipes but these recipes are identical (read: same variant-id). Then Bob will still build the right thing but in `bob ls` you will see only one of the `$BUSYBOX_CONFIG` dependencies and the other one is silently dropped.

> -> There is no obvious need to add `BUSYBOX_CONFIG` as `buildVar`. I think this should be at least documented somewhere...

There is no need to add `BUSYBOX_CONFIG` as `buildVar`. What may be documented is that identical dependencies with different names might "vanish". But I consider this an outlier because I have never encountered different recipes that were identical. That's why I have never added code to handle this case that I though will never happen in the first place.